// xijs/dist/index.esm.js
export { nanoid as uuid } from 'nanoid';
import _toConsumableArray from '@babel/runtime/helpers/esm/toConsumableArray';
import { parse } from 'qs';
import _regeneratorRuntime from '@babel/runtime/regenerator';
import _asyncToGenerator from '@babel/runtime/helpers/esm/asyncToGenerator';
import _createForOfIteratorHelper from '@babel/runtime/helpers/esm/createForOfIteratorHelper';
import _slicedToArray from '@babel/runtime/helpers/esm/slicedToArray';
import _typeof from '@babel/runtime/helpers/esm/typeof';
export { default as difference } from 'lodash.difference';
import _createClass from '@babel/runtime/helpers/esm/createClass';
import _classCallCheck from '@babel/runtime/helpers/esm/classCallCheck';

var parser = {
  FUNC_PREFIX: '__xfunc__',
  REG_PREFIX: '__xreg__',
  isArrOrObj: function isArrOrObj(obj) {
    return this.isArr(obj) || this.isObj(obj);
  },
  isArr: function isArr(obj) {
    return !!obj && Object.prototype.toString.call(obj) === '[object Array]';
  },
  isObj: function isObj(obj) {
    return !!obj && Object.prototype.toString.call(obj) === '[object Object]';
  },
  isRegExp: function isRegExp(obj) {
    return !!obj && Object.prototype.toString.call(obj) === '[object RegExp]';
  },
  isFunc: function isFunc(obj) {
    return typeof obj === 'function';
  },
  funcParse: function funcParse(obj) {
    var _this = this;

    var result;

    if (this.isArr(obj)) {
      result = [];
      result = obj.map(function (v) {
        /** 如果是函数类型, 则转换为字符串 */
        if (_this.isFunc(v)) {
          return "".concat(_this.FUNC_PREFIX).concat(v);
        }
        /** 如果类型为正则 */


        if (_this.isRegExp(v)) {
          return "".concat(_this.REG_PREFIX).concat(v);
        }
        /** 如果是数组或者对象 */


        if (_this.isArrOrObj(v)) {
          return _this.funcParse(v);
        }
        /** 基本类型 */


        return v;
      });
    }

    if (this.isObj(obj)) {
      result = {};

      for (var key in obj) {
        var v = obj[key];

        if (this.isFunc(v)) {
          result[key] = "".concat(this.FUNC_PREFIX).concat(v);
        } else if (this.isRegExp(v)) {
          /** 如果类型为正则 */
          result[key] = "".concat(this.REG_PREFIX).concat(v);
        } else if (this.isArrOrObj(v)) {
          result[key] = this.funcParse(v);
        } else result[key] = v;
      }
    }

    return result;
  },
  parse: function parse(jsonStr, error) {
    var _this2 = this;

    try {
      return JSON.parse(jsonStr, function (key, value) {
        if (value && typeof value === 'string') {
          var _value = value.indexOf(_this2.FUNC_PREFIX) > -1 ? new Function("return ".concat(value.replace(_this2.FUNC_PREFIX, '')))() : value.indexOf(_this2.REG_PREFIX) > -1 ? new Function("return ".concat(value.replace(_this2.REG_PREFIX, '')))() : value;

          return _value;
        }

        return value;
      });
    } catch (err) {
      error && error(err);
    }
  },
  stringify: function stringify(obj, replacer, space, error) {
    try {
      var _obj = obj;

      if (this.isRegExp(obj)) {
        /** 如果类型为正则 */
        _obj = "".concat(this.REG_PREFIX).concat(obj);
      }

      if (this.isFunc(obj)) {
        _obj = "".concat(this.FUNC_PREFIX).concat(obj);
      }

      if (this.isArrOrObj(obj)) {
        _obj = this.funcParse(obj);
      }

      return JSON.stringify(_obj, replacer, space);
    } catch (err) {
      error && error(err);
      return '';
    }
  },
  fastStringify: function fastStringify(obj, space, error) {
    var _this3 = this;

    try {
      return JSON.stringify(obj, function (k, v) {
        /** 如果类型为函数 */
        if (_this3.isFunc(v)) {
          return "".concat(_this3.FUNC_PREFIX).concat(v);
        }
        /** 如果类型为正则 */


        if (_this3.isRegExp(v)) {
          return "".concat(_this3.REG_PREFIX).concat(v);
        }

        return v;
      }, space);
    } catch (err) {
      error && error(err);
      return '';
    }
  },
  nativeStringify: JSON.stringify,
  nativeParse: JSON.parse
};

/**
 *  支持过期时间的localStorage库
 */
var store = {
  preId: 'xi-',
  timeSign: '|-door-|',
  status: {
    SUCCESS: 0,
    FAILURE: 1,
    OVERFLOW: 2,
    TIMEOUT: 3
  },
  storage: localStorage || window.localStorage,
  getKey: function getKey(key) {
    return this.preId + key;
  },
  set: function set(key, value, time, cb) {
    var _status = this.status.SUCCESS,
        _time;

    var _key = this.getKey(key); // 设置失效时间，未设置时间默认为一个月


    try {
      _time = time ? new Date(time).getTime() || time.getTime() : new Date().getTime() + 1000 * 60 * 60 * 24 * 31;
    } catch (e) {
      _time = new Date().getTime() + 1000 * 60 * 60 * 24 * 31;
    }

    try {
      this.storage.setItem(_key, _time + this.timeSign + value);
    } catch (e) {
      _status = this.status.OVERFLOW;
    }

    cb && cb.call(this, _status, _key, value);
  },
  get: function get(key, cb) {
    var _key = this.getKey(key),
        timeSignLen = this.timeSign.length,
        that = this;

    var status = this.status.SUCCESS,
        value = null,
        index,
        time,
        result;

    try {
      value = that.storage.getItem(_key);
    } catch (e) {
      result = {
        status: that.status.FAILURE,
        value: null
      };
      cb && cb.call(this, result.status, result.value);
      return result;
    }

    if (value) {
      index = value.indexOf(that.timeSign);
      time = +value.slice(0, index);

      if (time > new Date().getTime() || time == 0) {
        value = value.slice(index + timeSignLen);
      } else {
        value = null, status = that.status.TIMEOUT;
        that.remove(key);
      }
    } else {
      status = that.status.FAILURE;
    }

    result = {
      status: status,
      value: value
    };
    cb && cb.call(this, result.status, result.value);
    return result;
  },
  // 删除storage，如果删除成功，返回删除的内容
  remove: function remove(key, cb) {
    var _key = this.getKey(key);

    var status = this.status.FAILURE,
        value;

    try {
      value = this.storage.getItem(_key);
    } catch (e) {// dosomething
    }

    if (value) {
      try {
        this.storage.removeItem(_key);
        status = this.status.SUCCESS;
      } catch (e) {// dosomething
      }
    }

    cb && cb.call(this, status, status > 0 ? null : value ? value.slice(value.indexOf(this.timeSign) + this.timeSign.length) : null);
  }
};

/**
 * 对象转url参数
 * @param {*} data
 * @param {*} isPrefix
 */
var obj2url = function obj2url(data, isPrefix) {
  var prefix = isPrefix ? '?' : '';
  var _result = [];

  var _loop = function _loop(key) {
    var value = data[key]; // 去掉为空的参数

    if (['', undefined, null].includes(value)) {
      return "continue";
    }

    if (value.constructor === Array) {
      value.forEach(function (_value) {
        _result.push(encodeURIComponent(key) + '[]=' + encodeURIComponent(_value));
      });
    } else {
      _result.push(encodeURIComponent(key) + '=' + encodeURIComponent(value));
    }
  };

  for (var key in data) {
    var _ret = _loop(key);

    if (_ret === "continue") continue;
  }

  return _result.length ? prefix + _result.join('&') : '';
};

var IsPc = function IsPc() {
  var userAgentInfo = navigator.userAgent;
  var Agents = ['Android', 'iPhone', 'SymbianOS', 'Windows Phone', 'iPad', 'iPod'];
  var flag = true;

  for (var v = 0; v < Agents.length; v++) {
    if (userAgentInfo.indexOf(Agents[v]) > 0) {
      flag = false;
      break;
    }
  }

  return flag;
};

/**
 * 时间格式化
 * @param time
 * @param format
 * @returns
 */
function formatDate(time) {
  var format = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'YY-MM-DD hh:mm:ss';
  var date = new Date(time);
  var year = date.getFullYear(),
      month = date.getMonth() + 1,
      //月份是从0开始的
  day = date.getDate(),
      hour = date.getHours(),
      min = date.getMinutes(),
      sec = date.getSeconds();

  var preArr = _toConsumableArray(Array(10)).map(function (elem, index) {
    return '0' + index;
  });

  var newTime = format.replace(/YY/g, String(year)).replace(/MM/g, String(preArr[month] || month)).replace(/DD/g, String(preArr[day] || day)).replace(/hh/g, String(preArr[hour] || hour)).replace(/mm/g, String(preArr[min] || min)).replace(/ss/g, String(preArr[sec] || sec));
  return newTime;
}

function url2obj() {
  var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '?a=1&b=2&c=3';

  if (params.indexOf('?') > -1) {
    params = params.slice(1);
  }

  return parse(params);
}

/**
 *
 * @param fn
 * @param delay
 * @returns
 */
function throttle(fn, delay) {
  var flag = true;
  return function () {
    if (flag) {
      flag = false;
      fn.apply(void 0, arguments);
      setTimeout(function () {
        flag = true;
      }, delay);
    }
  };
}

/**
 *
 * @param fn 回调函数
 * @param wait 延迟时间
 * @returns
 */
function debounce(fn) {
  var wait = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 300;
  var timer = null;
  return function () {
    var _this = this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    if (timer) clearTimeout(timer);
    timer = setTimeout(function () {
      fn.apply(_this, args);
    }, wait);
  };
}

/**
 * 生成指定个数的随机字符串
 * @param n
 * @param str
 * @returns
 */
function randomStr(n) {
  var str = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'abcdefghigklmnopqrstuvexyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890_-+%$';
  var res = '';
  var len = str.length;

  for (var i = 0; i < n; i++) {
    res += str[Math.floor(Math.random() * len)];
  }

  return res;
}

/**
 *
 * @param str
 * @param char
 * @returns
 */
var charCount = function charCount(str, char) {
  return str.split(char).length - 1;
};

/**
 *
 * @param obj
 * @returns
 */
var isEmpty = function isEmpty(obj) {
  return Reflect.ownKeys(obj).length === 0 && obj.constructor === Object;
};

function isPhone(phone) {
  var myreg = /^[1][3,4,5,6,7,8,9][0-9]{9}$/;

  if (!myreg.test(phone)) {
    return false;
  } else {
    return true;
  }
}

/**
 *
 * @param milliseconds
 * @returns
 */
var sleep = /*#__PURE__*/function () {
  var _ref = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(milliseconds) {
    return _regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            return _context.abrupt("return", new Promise(function (resolve) {
              return setTimeout(resolve, milliseconds);
            }));

          case 1:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));

  return function sleep(_x) {
    return _ref.apply(this, arguments);
  };
}();

/**
 *
 * @param url
 * @returns
 */
var redirect = function redirect(url) {
  return location.href = url;
};

/**
 *
 * @param arr
 * @returns
 */
var shuffle = function shuffle(arr) {
  return arr.sort(function () {
    return 0.5 - Math.random();
  });
};

/**
 * 获取选中的文本
 * @returns
 */
var getSelection = function getSelection() {
  return window.getSelection().toString();
};

/**
 * 计算数组平均值
 * @returns
 */
var average = function average(arr) {
  return arr.reduce(function (a, b) {
    return a + b;
  }) / arr.length;
};

/**
 * 横线转驼峰命名
 * @param str
 * @returns
 */
function camelize(str) {
  var camelizeRE = /-(\w)/g;
  return str.replace(camelizeRE, function (_, c) {
    return c ? c.toUpperCase() : '';
  });
}

/**
 * 驼峰命名转横线命名
 * @param str
 * @returns
 */
function hyphenate(str) {
  var flag = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '-';
  var hyphenateRE = /\B([A-Z])/g;
  return str.replace(hyphenateRE, "".concat(flag, "$1")).toLowerCase();
}

/**
 * 获取数据类型，返回结果为Number、String、Object、Array等
 * @param value
 * @returns
 */
function getRawType(value) {
  return Object.prototype.toString.call(value).slice(8, -1);
}

/**
 * 获取当前运行环境,返回运行环境的集合判断
 * @returns
 */
function getRuntimeEnv() {
  var inBrowser = typeof window !== 'undefined';
  var inWeex = typeof window.WXEnvironment !== 'undefined' && !!window.WXEnvironment.platform;
  var weexPlatform = inWeex && window.WXEnvironment.platform.toLowerCase(); //浏览器 UA 判断

  var UA = inBrowser && window.navigator.userAgent.toLowerCase();
  var isIE = UA && /msie|trident/.test(UA);
  var isIE9 = UA && UA.indexOf('msie 9.0') > 0;
  var isEdge = UA && UA.indexOf('edge/') > 0;
  var isAndroid = UA && UA.indexOf('android') > 0 || weexPlatform === 'android';
  var isIOS = UA && /iphone|ipad|ipod|ios/.test(UA) || weexPlatform === 'ios';
  var isChrome = UA && /chrome\/\d+/.test(UA) && !isEdge;
  return {
    isWeex: inWeex,
    isIE: isIE,
    isIE9: isIE9,
    isEdge: isEdge,
    isAndroid: isAndroid,
    isIOS: isIOS,
    isChrome: isChrome
  };
}

/**
 * 生成重复字符串
 * @param str 需要重复的字符串
 * @param n 重复次数
 * @returns
 */
function repeat(str, n) {
  var res = '';

  while (n) {
    if (n % 2 === 1) {
      res += str;
    }

    if (n > 1) {
      str += str;
    }

    n >>= 1;
  }

  return res;
}

/**
 * 将扁平数组转换成树结构
 * @param list 扁平数组
 * @param options 树解析map
 * @returns
 */
function transformTree() {
  var list = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {
    keyField: 'id',
    childField: 'children',
    parentField: 'pid'
  };
  var keyField = options.keyField,
      childField = options.childField,
      parentField = options.parentField;
  var tree = [];
  var record = {};

  for (var i = 0, len = list.length; i < len; i++) {
    var item = list[i];
    var id = item[keyField];

    if (!id) {
      continue;
    }

    if (record[id]) {
      item[childField] = record[id];
    } else {
      item[childField] = record[id] = [];
    }

    if (item[parentField]) {
      var parentId = item[parentField];

      if (!record[parentId]) {
        record[parentId] = [];
      }

      record[parentId].push(item);
    } else {
      tree.push(item);
    }
  }

  return tree;
}

/**
 * 返回区域内随机数
 * @param lower 下限
 * @param upper 上限
 * @returns
 */
function random(lower, upper) {
  lower = +lower || 0;
  upper = +upper || 0;
  return Math.random() * (upper - lower) + lower;
}

/**
 * 深拷贝
 * @returns
 */
function cloneDeep(obj) {
  if (obj === null) {
    return null;
  }

  if (obj instanceof Set) {
    return new Set(_toConsumableArray(obj));
  }

  if (obj instanceof Map) {
    return new Map(_toConsumableArray(obj));
  }

  if (obj instanceof WeakMap) {
    var weakMap = new WeakMap();
    weakMap = obj;
    return weakMap;
  }

  if (obj instanceof WeakSet) {
    var weakSet = new WeakSet();
    weakSet = obj;
    return weakSet;
  }

  if (obj instanceof RegExp) {
    return new RegExp(obj);
  }

  if (typeof obj === 'undefined') {
    return undefined;
  }

  if (Array.isArray(obj)) {
    return obj.map(cloneDeep);
  }

  if (obj instanceof Date) {
    return new Date(obj.getTime());
  }

  if (_typeof(obj) !== 'object') {
    return obj;
  }

  var newObj = {};

  for (var _i = 0, _Object$entries = Object.entries(obj); _i < _Object$entries.length; _i++) {
    var _Object$entries$_i = _slicedToArray(_Object$entries[_i], 2),
        key = _Object$entries$_i[0],
        value = _Object$entries$_i[1];

    newObj[key] = cloneDeep(value);
  }

  var symbolkeys = Object.getOwnPropertySymbols(obj);

  var _iterator = _createForOfIteratorHelper(symbolkeys),
      _step;

  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var _key = _step.value;
      newObj[_key] = cloneDeep(obj[_key]);
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }

  return newObj;
}

/**
 * base64编码和解码
 * @param str
 * @returns
 */
var base64 = {
  _keyStr: 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=',
  encode: function encode(e) {
    var t = '';
    var n, r, i, s, o, u, a;
    var f = 0;
    e = base64._utf8_encode(e);

    while (f < e.length) {
      n = e.charCodeAt(f++);
      r = e.charCodeAt(f++);
      i = e.charCodeAt(f++);
      s = n >> 2;
      o = (n & 3) << 4 | r >> 4;
      u = (r & 15) << 2 | i >> 6;
      a = i & 63;

      if (isNaN(r)) {
        u = a = 64;
      } else if (isNaN(i)) {
        a = 64;
      }

      t = t + this._keyStr.charAt(s) + this._keyStr.charAt(o) + this._keyStr.charAt(u) + this._keyStr.charAt(a);
    }

    return t;
  },
  decode: function decode(e) {
    var t = '';
    var n, r, i;
    var s, o, u, a;
    var f = 0;
    e = e.replace(/[^A-Za-z0-9+/=]/g, '');

    while (f < e.length) {
      s = this._keyStr.indexOf(e.charAt(f++));
      o = this._keyStr.indexOf(e.charAt(f++));
      u = this._keyStr.indexOf(e.charAt(f++));
      a = this._keyStr.indexOf(e.charAt(f++));
      n = s << 2 | o >> 4;
      r = (o & 15) << 4 | u >> 2;
      i = (u & 3) << 6 | a;
      t = t + String.fromCharCode(n);

      if (u != 64) {
        t = t + String.fromCharCode(r);
      }

      if (a != 64) {
        t = t + String.fromCharCode(i);
      }
    }

    t = base64._utf8_decode(t);
    return t;
  },
  _utf8_encode: function _utf8_encode(e) {
    e = e.replace(/rn/g, 'n');
    var t = '';

    for (var n = 0; n < e.length; n++) {
      var r = e.charCodeAt(n);

      if (r < 128) {
        t += String.fromCharCode(r);
      } else if (r > 127 && r < 2048) {
        t += String.fromCharCode(r >> 6 | 192);
        t += String.fromCharCode(r & 63 | 128);
      } else {
        t += String.fromCharCode(r >> 12 | 224);
        t += String.fromCharCode(r >> 6 & 63 | 128);
        t += String.fromCharCode(r & 63 | 128);
      }
    }

    return t;
  },
  _utf8_decode: function _utf8_decode(e) {
    var t = '',
        n = 0,
        r = 0,
        c2 = 0,
        c3 = 0;

    while (n < e.length) {
      r = e.charCodeAt(n);

      if (r < 128) {
        t += String.fromCharCode(r);
        n++;
      } else if (r > 191 && r < 224) {
        c2 = e.charCodeAt(n + 1);
        t += String.fromCharCode((r & 31) << 6 | c2 & 63);
        n += 2;
      } else {
        c2 = e.charCodeAt(n + 1);
        c3 = e.charCodeAt(n + 2);
        t += String.fromCharCode((r & 15) << 12 | (c2 & 63) << 6 | c3 & 63);
        n += 3;
      }
    }

    return t;
  }
};

/**
 * 紧凑型驼峰命名转横线命名
 * @param str
 * @returns
 */
function hyCompact(str) {
  var flag = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '-';
  var temp = str.replace(/[A-Z]{1,2}/g, function (match) {
    return flag + match.toLowerCase();
  });

  if (temp.slice(0, 1) === flag) {
    //如果首字母是大写，执行replace时会多一个-，这里需要去掉
    temp = temp.slice(1);
  }

  return temp;
}

/**
 *
 * @param any
 * @returns
 */
var isArray = function isArray(obj) {
  return Object.prototype.toString.call(obj) === '[object Array]';
};

// 压缩前将file转换成img对象
function readImg(file) {
  return new Promise(function (resolve, reject) {
    var img = new Image();
    var reader = new FileReader();

    reader.onload = function (e) {
      img.src = e.target.result;
    };

    reader.onerror = function (e) {
      reject(e);
    };

    reader.readAsDataURL(file);

    img.onload = function () {
      resolve(img);
    };

    img.onerror = function (e) {
      reject(e);
    };
  });
}

// 将rgba字符串对象转化为rgba对象
function rgba2obj() {
  var rgba = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
  var reg = /rgba\((\d+),(\d+),(\d+),(\d?\.?\d+)\)/g;
  var rgbaObj = {
    r: 0,
    g: 0,
    b: 0,
    a: 0
  };
  rgba.replace(reg, function (_m, r, g, b, a) {
    rgbaObj = {
      r: +r,
      g: +g,
      b: +b,
      a: +a
    };
    return rgba;
  });
  return rgbaObj;
}

// 将hex字符串转化为rgba及其对象
function hex2rgba(ahex) {
  var alpha = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 1;
  var BASE = 16;
  var HEX_REGEX = /^#?[a-fA-F0-9]+$/;
  var HEX_SHORTHAND_LENGTH = 3;
  var HEX_LENGTH = 6;
  var hex = ahex;

  if (!HEX_REGEX.test(hex)) {
    throw Error('hex2rgba: 第一个参数为非法的hex字符串');
  } // 剔除#


  if (hex[0] === '#') {
    hex = hex.slice(1);
  } // 如果是hex简写，如#fff, 统一转换为#ffffff


  if (hex.length === HEX_SHORTHAND_LENGTH) {
    hex = hex.split('');
    hex.splice(2, 0, hex[2]);
    hex.splice(1, 0, hex[1]);
    hex.splice(0, 0, hex[0]);
    hex = hex.join('');
  }

  if (hex.length !== HEX_LENGTH) {
    throw Error('hex2rgba: 非法的hex长度');
  } // hex 转换为 rgb


  var values = [parseInt(hex.slice(0, 2), BASE), parseInt(hex.slice(2, 4), BASE), parseInt(hex.slice(4, 6), BASE)];
  alpha = typeof alpha === 'number' ? alpha : parseFloat(alpha);

  if (alpha >= 0 && alpha <= 1) {
    values.push(alpha);
  } else {
    values.push(1);
  }

  return {
    v: 'rgba(' + values.join(',') + ')',
    o: {
      r: values[0],
      g: values[1],
      b: values[2],
      a: values[3]
    }
  };
}

/**
 * 数字千分位展示并显示n位小数
 * @param {(number | string)} 需要格式化的值
 * @param {number} [precision] 保留几位小数，不传小数不处理，不够会填充0
 * @return {(number | string)} 返回0或者格式化的值
 */
var formatNumber = function formatNumber(num, precision) {
  var parts; // 判断是否为数字

  if (!isNaN(parseFloat(num)) && isFinite(num)) {
    num = Number(num); // 处理小数点位数

    num = (typeof precision !== 'undefined' ? num.toFixed(precision) : num).toString(); // 分离数字的小数部分和整数部分

    parts = num.split('.');
    parts[0] = parts[0].toString().replace(/(\d)(?=(\d{3})+(?!\d))/g, '$1,');
    return parts.join('.');
  }

  return 0;
};

// 判断是否是有效的数字
var judgeIsNumber = function judgeIsNumber(value) {
  return typeof value === 'number' && !isNaN(value) && Math.abs(value) !== Infinity;
};
/**
 * 将值转换为百分数
 * @param value 任何值
 * @param precision 小数位
 */


function formatPercent(value) {
  var precision = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 2;
  var result = '0';

  if (judgeIsNumber(value) && value !== 0) {
    result = value % 1 === 0 ? "".concat(value * 100) : (value * 100).toFixed(precision);
  }

  return "".concat(result, "%");
}

/**
 * @param center 表示圆心的 x, y 坐标
 * @param r 表示圆的半径
 * @returns 返回圆内的任意坐标
 */
function coordinatesInCircle(center) {
  var r = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 1;

  if (center && Array.isArray(center) && center.length === 2) {
    var _center = _slicedToArray(center, 2),
        xAxis = _center[0],
        yAxis = _center[1],
        xValueRange = [xAxis - r, xAxis + r],
        yValueRange = [yAxis - r, yAxis + r];

    var pointX, pointY;

    while (true) {
      // 生成大于等于 xValueRange[0]，小于等于 xValueRange[1] 的 x 坐标
      pointX = Math.floor(Math.random() * (xValueRange[1] - xValueRange[0] + 1)) + xValueRange[0]; // 生成大于等于 yValueRange[0]，小于等于 yValueRange[1] 的 y 坐标

      pointY = Math.floor(Math.random() * (yValueRange[1] - yValueRange[0] + 1)) + yValueRange[0]; // 距离圆心的长度

      var sumOfSquares = Math.pow(pointX - xAxis, 2) + Math.pow(pointY - yAxis, 2);
      var distanceToCircle = Math.sqrt(sumOfSquares);

      if (distanceToCircle <= r) {
        return [pointX, pointY];
      }
    }
  } else {
    throw new Error('Please enter correct parameters.');
  }
}

/**
 * @param startPos 左上顶点坐标
 * @param w 矩形的宽
 * @param h 矩形的高
 * @returns 返回矩形内的任意坐标
 */
function coordinatesInRect(startPos, w, h) {
  if (startPos && Array.isArray(startPos) && startPos.length === 2 && typeof w === 'number' && typeof h === 'number') {
    var _startPos = _slicedToArray(startPos, 2),
        x0 = _startPos[0],
        y0 = _startPos[1]; // 生成大于等于 x0，小于等于 x0 + w 的 x 坐标


    var pointX = Math.floor(Math.random() * (w + 1)) + x0; // 生成大于等于 y0，小于等于 y0 + h 的 y 坐标

    var pointY = Math.floor(Math.random() * (h + 1)) + y0;
    return [pointX, pointY];
  } else {
    throw new Error('Please enter correct parameters.');
  }
}

/**
 * @param circle 描述圆形信息的对象，包含 center 属性和 r 属性，center 表示圆心的坐标，r 表示圆的半径。
 * @param x 该点的 x 轴坐标
 * @param y 该点的 y 轴坐标
 * @returns 如果该点在圆内，返回 true，否则返回 false。
 */
function judgePointInCircle(circle, x, y) {
  if (circle && circle.center && circle.center.length === 2 && typeof circle.r === 'number' && typeof x === 'number' && typeof y === 'number') {
    var center = circle.center,
        r = circle.r;

    var _center = _slicedToArray(center, 2),
        centerX = _center[0],
        centerY = _center[1];

    var sumOfSquares = Math.pow(x - centerX, 2) + Math.pow(y - centerY, 2);
    var distanceToCircle = Math.sqrt(sumOfSquares);
    return distanceToCircle <= r;
  } else {
    throw new Error('Please enter correct parameters.');
  }
}

/**
 * 对象数组、普通数组冒泡排序
 * @param obj
 * @param arrKey
 * @param orderby
 * @returns
 * 对象数组使用  bubbleSort(arr,key,orderby)
 * 普通数组使用  bubbleSort(arr) bubbleSort(arr,'',orderby)
 */

function bubbleSort(obj, arrKey) {
  var orderby = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'asc';
  var result = obj;
  var temp = null;

  if (isArray(obj)) {
    if (obj.length > 1) {
      for (var i = 0; i < result.length; i++) {
        for (var j = 0; j < i; j++) {
          var resultfiled_i = result[i][arrKey] ? result[i][arrKey] : result[i],
              resultfiled_j = result[j][arrKey] ? result[j][arrKey] : result[j]; // 降序

          if (orderby === 'desc') {
            if (parseInt(resultfiled_i) > parseInt(resultfiled_j)) {
              temp = result[j];
              result[j] = result[i];
              result[i] = temp;
            } // 升序

          } else {
            if (parseInt(resultfiled_i) < parseInt(resultfiled_j)) {
              temp = result[j];
              result[j] = result[i];
              result[i] = temp;
            }
          }
        }
      }
    }

    return result;
  } else throw Error('数据格式错误');
}

/**
 * 对象数组、普通数组快速排序
 * @param arr
 * @param arrKey
 * @returns
 */

function quickSort(arr, key) {
  if (isArray(arr)) {
    var result = arr;

    if (result.length <= 1) {
      return result;
    }

    var pivotIndex = Math.floor(result.length / 2);
    var pivot = result[pivotIndex];
    var left = [];
    var right = [];

    for (var i = 0; i < result.length; i++) {
      if (i === pivotIndex) {
        continue;
      }

      if ((key ? result[i][key] : result[i]) <= (key ? pivot[key] : pivot)) {
        left.push(result[i]);
      } else {
        right.push(result[i]);
      }
    }

    var resultArr = [].concat(_toConsumableArray(quickSort(left, key)), [pivot], _toConsumableArray(quickSort(right, key)));
    return resultArr;
  } else throw Error('数据格式错误');
}

/**
 *
 * @param n 需要计算阶乘的数字
 * @returns 若阶乘结果大于 Number.MAX_SAFE_INTEGER，则返回 bigint 数据类型
 */
function factorial(n) {
  var res = 1,
      i = 1; // 大于等于19的阶乘结果超出了 Number.MAX_SAFE_INTEGER，需要 bigint 数据类型存储数据

  if (n >= 19) {
    res = BigInt(1);
    i = BigInt(1);
  }

  while (i <= n) {
    res = res * i;
    i++;
  }

  return res;
}

/**
 *
 * @param n
 * @returns 返回第n个斐波那契数
 */
function fibonacci(n) {
  if (n === 0) return 0;
  if (n === 1) return 1;
  var n2 = 0,
      n1 = 1,
      res = 0;

  for (var i = 2; i <= n; i++) {
    res = n1 + n2;
    n2 = n1;
    n1 = res;
  }

  return res;
}

/**
 *
 * @param arr 求和数组
 * @returns 求和的结果
 */
function sum(arr) {
  return arr.reduce(function (pre, cur) {
    return pre + cur;
  }, 0);
}

/**
 *
 * @param email
 * @returns
 */
var isEmail = function isEmail(email) {
  var emailRegex = /\S+@\S+\.\S+/;
  return emailRegex.test(email);
};

/**
 *
 * @param idCard
 * @returns
 */
// 身份证校验
var isIdCard = function isIdCard(idCard) {
  // areaCode:地区码  checkCode：末尾校验码
  var areaCode = [11, 12, 13, 14, 15, 21, 22, 23, 31, 32, 33, 34, 35, 36, 37, 41, 42, 43, 44, 45, 46, 50, 51, 52, 53, 54, 61, 62, 63, 64, 65, 71, 81, 82, 91];
  var checkCode = ['1', '0', 'X', '9', '8', '7', '6', '5', '4', '3', '2']; // 15位身份证号码

  if (idCard.length === 15) {
    // 判断地区码
    var idCardAreaCode = parseInt(idCard.substr(0, 2));

    if (areaCode.indexOf(idCardAreaCode) === -1) {
      return false;
    } // 判断时间


    var borthYear = parseInt(idCard.substr(6, 2)) + 1900;
    var isRunNian = borthYear % 400 === 0 || borthYear % 100 !== 0 && borthYear % 4 === 0;
    var regStr = /^[1-9][0-9]{5}[0-9]{2}((01|03|05|07|08|10|12)(0[1-9]|[1-2][0-9]|3[0-1])|(04|06|09|11)(0[1-9]|[1-2][0-9]|30)|02(0[1-9]|1[0-9]|2[0-8]))[0-9]{3}$/;

    if (isRunNian) {
      regStr = /^[1-9][0-9]{5}[0-9]{2}((01|03|05|07|08|10|12)(0[1-9]|[1-2][0-9]|3[0-1])|(04|06|09|11)(0[1-9]|[1-2][0-9]|30)|02(0[1-9]|[1-2][0-9]))[0-9]{3}$/;
    }

    if (!idCard.match(regStr)) {
      return false;
    }

    return true; // 18位身份证号码
  } else if (idCard.length === 18) {
    var _idCardAreaCode = parseInt(idCard.substr(0, 2));

    if (areaCode.indexOf(_idCardAreaCode) === -1) {
      return false;
    }

    var _borthYear = parseInt(idCard.substr(6, 4));

    var _isRunNian = _borthYear % 400 === 0 || _borthYear % 100 !== 0 && _borthYear % 4 === 0;

    var _regStr = /^[1-9][0-9]{5}(19|20)[0-9]{2}((01|03|05|07|08|10|12)(0[1-9]|[1-2][0-9]|3[0-1])|(04|06|09|11)(0[1-9]|[1-2][0-9]|30)|02(0[1-9]|1[0-9]|2[0-8]))[0-9]{3}[0-9Xx]$/;

    if (_isRunNian) {
      _regStr = /^[1-9][0-9]{5}(19|20)[0-9]{2}((01|03|05|07|08|10|12)(0[1-9]|[1-2][0-9]|3[0-1])|(04|06|09|11)(0[1-9]|[1-2][0-9]|30)|02(0[1-9]|[1-2][0-9]))[0-9]{3}[0-9Xx]$/;
    }

    if (!idCard.match(_regStr)) {
      return false;
    } // 判断最后一位


    var sumIdCard = (parseInt(idCard.substr(0, 1)) + parseInt(idCard.substr(10, 1))) * 7 + (parseInt(idCard.substr(1, 1)) + parseInt(idCard.substr(11, 1))) * 9 + (parseInt(idCard.substr(2, 1)) + parseInt(idCard.substr(12, 1))) * 10 + (parseInt(idCard.substr(3, 1)) + parseInt(idCard.substr(13, 1))) * 5 + (parseInt(idCard.substr(4, 1)) + parseInt(idCard.substr(14, 1))) * 8 + (parseInt(idCard.substr(5, 1)) + parseInt(idCard.substr(15, 1))) * 4 + (parseInt(idCard.substr(6, 1)) + parseInt(idCard.substr(16, 1))) * 2 + parseInt(idCard.substr(7, 1)) * 1 + parseInt(idCard.substr(8, 1)) * 6 + parseInt(idCard.substr(9, 1)) * 3;
    var modNum = checkCode[sumIdCard % 11];

    if (modNum !== idCard.substr(17, 1).toUpperCase()) {
      return false;
    }

    return true;
  } else {
    return false;
  }
};

/**
 *
 * @param linkList
 * @returns
 */
var linkListToArray = function linkListToArray(linkList) {
  var arr = [];
  var header = linkList;

  while (header) {
    arr.push(header.value);
    header = header.next;
  }

  return arr;
};

var lang = {
  /**
   *
   * @param str
   * @returns
   */
  isCn: function isCn(str) {
    var chineseRegex = /^[\u4e00-\u9fa5]+$/;
    return chineseRegex.test(str);
  },

  /**
   *
   * @param str
   * @returns
   */
  isEn: function isEn(str) {
    var englishRegex = /^[A-Za-z]+$/;
    return englishRegex.test(str);
  }
};

/**
 * @param tree 树结构
 * @param childField 每个节点中子树的字段名
 * @returns 扁平数组
 */
function transformArray() {
  var tree = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
  var childField = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'children';
  // 层序遍历
  var treeNodes = [];
  var res = [];

  var _iterator = _createForOfIteratorHelper(tree),
      _step;

  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var rootNode = _step.value;
      treeNodes.push(rootNode);
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }

  while (treeNodes.length > 0) {
    var node = treeNodes.shift();

    if (node[childField] && node[childField].length > 0) {
      var _iterator2 = _createForOfIteratorHelper(node[childField]),
          _step2;

      try {
        for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
          var childNode = _step2.value;
          treeNodes.push(childNode);
        }
      } catch (err) {
        _iterator2.e(err);
      } finally {
        _iterator2.f();
      }
    }

    delete node[childField];
    res.push(node);
  }

  return res;
}

var arrayToListNode = function arrayToListNode(arr) {
  var len = arr.length;

  if (!len) {
    return undefined;
  }

  var node = {
    data: arr[len - 1]
  };

  for (var i = len - 2; i >= 0; i--) {
    node = {
      data: arr[i],
      next: node
    };
  }

  return node;
};

/*
 * @Author: bigFace2019 599069310@qq.com
 * @Date: 2023-04-11 19:40:55
 * @LastEditors: bigFace2019 599069310@qq.com
 * @LastEditTime: 2023-04-11 20:22:32
 * @FilePath: \xijs\src\getDom\index.ts
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
 */
function $(selector) {
  var nodes = document.querySelectorAll(selector);

  if (nodes.length === 0) {
    return undefined;
  }

  if (nodes.length === 1) {
    return nodes[0];
  }

  return nodes;
}

/*
 * @Author: bigFace2019 599069310@qq.com
 * @Date: 2023-04-11 19:40:55
 * @LastEditors: bigFace2019 599069310@qq.com
 * @LastEditTime: 2023-04-13 20:31:22
 * @FilePath: \xijs\src\getDom\index.ts
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
 */

function getDomText(selector) {
  var doms = $(selector); //1.是undefined

  if (!doms) {
    return '';
  } //1.是nodeList


  if (Reflect.has(doms, 'length')) {
    var res = '';
    Array.from(doms).forEach(function (item) {
      res += item.textContent;
    });
    return res;
  } //3是element


  return doms.textContent;
}

/*
 * @Author: bigFace2019 599069310@qq.com
 * @Date: 2023-04-11 19:40:55
 * @LastEditors: bigFace2019 599069310@qq.com
 * @LastEditTime: 2023-04-14 19:23:31
 * @FilePath: \xijs\src\getDom\index.ts
 * @Description: 获取dom节点相对于文档定位。
 */
function getDomPageXY(selector) {
  var element = document.querySelector(selector);

  if (element) {
    var _element$getBoundingC = element.getBoundingClientRect(),
        top = _element$getBoundingC.top,
        left = _element$getBoundingC.left; //top和left加上滚动条
    // For scrollX


    var scrollLeft = document.documentElement.scrollLeft; // For scrollY

    var scrollTop = document.documentElement.scrollTop;
    return [left + scrollLeft, top + scrollTop];
  }

  return undefined;
}

/*
 * @Author: bigFace2019 599069310@qq.com
 * @Date: 2023-04-11 19:40:55
 * @LastEditors: bigFace2019 599069310@qq.com
 * @LastEditTime: 2023-04-14 19:23:45
 * @FilePath: \xijs\src\getDom\index.ts
 * @Description: 获取dom节点相对于screen定位。
 */
function getDomScreenXY(selector) {
  var element = document.querySelector(selector);

  if (element) {
    var _element$getBoundingC = element.getBoundingClientRect(),
        top = _element$getBoundingC.top,
        left = _element$getBoundingC.left;

    return [left, top];
  }

  return undefined;
}

// xss 注入，对标签进行转义
var xss = function xss(str) {
  return str.replace(/[&<>'"]/g, function (tag) {
    return {
      '&': '&amp;',
      '<': '&lt;',
      '>': '&gt;',
      "'": '&#39;',
      '"': '&quot;'
    }[tag] || tag;
  });
};

var dateTypeMap = {
  year: function year(date, num) {
    return date.setFullYear(date.getFullYear() + num);
  },
  month: function month(date, num) {
    return date.setMonth(date.getMonth() + num);
  },
  day: function day(date, num) {
    return date.setDate(date.getDate() + num);
  },
  hour: function hour(date, num) {
    return date.setHours(date.getHours() + num);
  },
  minute: function minute(date, num) {
    return date.setMinutes(date.getMinutes() + num);
  },
  second: function second(date, num) {
    return date.setSeconds(date.getSeconds() + num);
  },
  milliseconds: function milliseconds(date, num) {
    return date.setMilliseconds(date.getMilliseconds() + num);
  }
};
/**
 * 对日期进行计算，比如计算前一天，前一个月，后一天，后一个月等等
 * @param dateStr: 需要计算的日期字符串或者 Date
 * @param num：需要加上或者减去的天数（小时，月，年，分钟，秒等）
 * @param flag：对 年月日时分秒 那个进行计算
 * @param format：对 计算后的日期进行格式化的格式，默认使用 YY-MM-DD hh:mm:ss。比如：2020-01-02 08:00:00
 */

var dateCalculate = function dateCalculate(dateStr, num, flag, format) {
  if (dateTypeMap[flag]) {
    var date = new Date(dateStr);
    var result = dateTypeMap[flag](date, num);
    return formatDate(result, format);
  } else {
    throw Error('请输入正确的计算参数(仅支持 年、月、日、时、分、秒、毫秒)');
  }
};

/**
 * 计算两个时间的差
 * @param start：开始时间
 * @param end：结束时间
 */
var timeSub = function timeSub(start, end) {
  var startDate = new Date(start);
  var endDate = new Date(end);
  var date = endDate.getTime() - startDate.getTime();
  var misSecond = Math.floor(date);
  return misSecond;
};

/*
 * @Author: bigFace2019 599069310@qq.com
 * @Date: 2023-04-11 19:40:55
 * @LastEditors: bigFace2019 599069310@qq.com
 * @LastEditTime: 2023-04-20 20:26:16
 * @FilePath: \xijs\src\getDom\index.ts
 * @Description:获取随机颜色'hsl'|'hsla'|'rgb'|'rgba'|16，默认是rgb格式。
 */

function genRandomColor(type) {
  var genColorObj = {
    hsl: 'hsl(' + Math.round(Math.random() * 360) + ',' + Math.round(Math.random() * 100) + '%,' + Math.round(Math.random() * 100) + '%)',
    hsla: 'hsla(' + Math.round(Math.random() * 360) + ',' + Math.round(Math.random() * 100) + '%,' + Math.round(Math.random() * 100) + '%,' + (Math.random() * 10 / 10).toFixed(1) + ')',
    rgb: 'rgb(' + Math.round(Math.random() * 255) + ',' + Math.round(Math.random() * 255) + ',' + Math.round(Math.random() * 255) + ')',
    rgba: 'rgba(' + Math.round(Math.random() * 255) + ',' + Math.round(Math.random() * 255) + ',' + Math.round(Math.random() * 255) + ',' + (Math.random() * 10 / 10).toFixed(1) + ')',
    16: '#' + Math.random().toString(16).substr(2, 6)
  };

  if (!type) {
    return genColorObj['rgb'];
  }

  if (['hsl', 'hsla', 'rgb', 'rgba', '16', 16].includes(type)) {
    return genColorObj[type];
  }
}

/**
 *
 * @param Regexp: 需要校验的正则表达式
 * @param text：需要匹配的内容
 */
var regex = function regex(regexp, text) {
  if (regexpMap[regexp]) {
    return regexpMap[regexp](text);
  } else {
    throw Error('请输入正确的正则类型');
  }
};
/**
 * 正则表达式支持的列表
 */


var regexpMap = {
  // 邮箱 的正则表达式
  emailRegexp: function emailRegexp(text) {
    return /^\w+([-+.]\w+)*@\w+([-.]\w+)*\.\w+([-.]\w+)*$/.test(text);
  },
  // 手机号码 的正则表达式
  phoneRegexp: function phoneRegexp(text) {
    return /^(13[0-9]|14[5|7]|15[0|1|2|3|4|5|6|7|8|9]|18[0|1|2|3|5|6|7|8|9])\d{8}$/.test(text);
  },
  // 国内电话号码 的正则表达式
  fixPhone: function fixPhone(text) {
    return /\d{3}-\d{8}|\d{4}-\d{7}/.test(text);
  },
  // 身份证号(15位、18位数字) 的正则表达式
  idRegexp: function idRegexp(text) {
    return /(^\d{15}$)|(^\d{18}$)|(^\d{17}(\d|X|x)$)/.test(text);
  },
  // IPv4 的正则表达式
  ipv4Regexp: function ipv4Regexp(text) {
    return /((2(5[0-5]|[0-4]\d))|[0-1]?\d{1,2})(\.((2(5[0-5]|[0-4]\d))|[0-1]?\d{1,2})){3}/.test(text);
  },
  // 是否包含中文 的正则表达式
  chineseRegexp: function chineseRegexp(text) {
    return /[\u4E00-\u9FA5]{2,4}$/.test(text);
  },
  // 匹配qq号（腾讯QQ号从10000开始） 的正则表达式
  qqRegexp: function qqRegexp(text) {
    return /[1-9][0-9]{4,}/.test(text);
  },
  // 匹配中国邮政编码 的正则表达式
  chinesePostcodeRegexp: function chinesePostcodeRegexp(text) {
    return /[1-9]\d{5}(?!d)/.test(text);
  },
  // 匹配英国邮政编码 的正则表达式
  usPostcodeRegexp: function usPostcodeRegexp(text) {
    return /^\d{5}(-\d{4})?$/.test(text);
  },
  // 匹配美国邮政编码 的正则表达式
  ukPostcodeRegexp: function ukPostcodeRegexp(text) {
    return /^[A-Za-z]{1,2}\d[A-Za-z\d]?\s*\d[A-Za-z]{2}$/.test(text);
  },

  /**
   * 车架号 的正则表达式
   * 由大写字母和数字组成，长度17位；
   * 字母不会出现O、Q、I三个字母；
   * 第9位只能是【0-9】的数字和字母X;
   * 第13-17位只能是数字；
   */
  carNumberRegexp: function carNumberRegexp(text) {
    return /^[A-HJ-NPR-Z\d]{8}[X\d][A-HJ-NPR-Z\d]{3}\d{5}$/.test(text);
  },
  // 帐号是否合法(字母开头，允许5-16位，允许字母数字下划线)： 正则表达式
  accountRegexp: function accountRegexp(text) {
    return /^[a-zA-Z][a-zA-Z0-9_]{4,15}$/.test(text);
  },
  // rgb 颜色 正则表达式
  rgbRegexp: function rgbRegexp(text) {
    return /^rgb\(\s*(\d{1,3})\s*,\s*(\d{1,3})\s*,\s*(\d{1,3})\s*\)$/.test(text);
  },
  // rgba 颜色 正则表达式
  rgbaRegexp: function rgbaRegexp(text) {
    return /^rgba\(\s*(\d{1,3})\s*,\s*(\d{1,3})\s*,\s*(\d{1,3})\s*,\s*([01](\.\d+)?)\s*\)$/.test(text);
  },
  // 十六进制颜色代码 正则表达式
  numberRegexp: function numberRegexp(text) {
    return /^#(?:[0-9a-fa-f]{3}){1,2}$/.test(text);
  },
  // 匹配 hsl 值 正则表达式
  hslRegexp: function hslRegexp(text) {
    return /^hsl\(\s*(\d{1,3})\s*,\s*(\d{1,3})%\s*,\s*(\d{1,3})%\s*\)$/.test(text);
  },
  // 匹配 hsla 值 正则表达式
  hslaRegexp: function hslaRegexp(text) {
    return /^hsla\(\s*(\d{1,3})\s*,\s*(\d{1,3})%\s*,\s*(\d{1,3})%\s*,\s*([01](\.\d+)?)\s*\)$/.test(text);
  }
};

/**
 *
 * @param text 下载文本
 * @param fileName 文件名
 */
function downloadFileWithText(text) {
  var fileName = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : text.slice(0, 5);
  var eleLink = document.createElement('a');
  eleLink.download = fileName;
  var blob = new Blob([text]);
  var blobUrl = URL.createObjectURL(blob);
  eleLink.href = blobUrl;
  eleLink.click();
  eleLink.remove();
  URL.revokeObjectURL(blobUrl);
}

/**
 *
 * @param url 下载链接
 * @param fileName 文件名
 */
function downloadFileWithUrl(url) {
  var fileName = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';

  if (!fileName) {
    var urlArr = url.split('/');
    fileName = urlArr[urlArr.length - 1];
  }

  fetch(url, {
    headers: new Headers({
      Origin: location.origin
    }),
    mode: 'no-cors'
  }).then( /*#__PURE__*/function () {
    var _ref = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(res) {
      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              return _context.abrupt("return", res.blob());

            case 1:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));

    return function (_x) {
      return _ref.apply(this, arguments);
    };
  }()).then(function (data) {
    var reader = new FileReader();
    reader.readAsDataURL(data);

    reader.onload = function () {
      var eleLink = document.createElement('a');
      eleLink.download = fileName;
      var blobUrl = reader.result;
      eleLink.href = blobUrl;
      eleLink.click();
      eleLink.remove();
    };
  });
}

/**
 * 计算传入的时间和当前时间的差值，比如计算几秒前、几分前、几秒后、几分后等等
 * @param timestamp 需要计算的时间戳
 */
var timeCutStr = function timeCutStr(time) {
  var newTime = time;

  if (typeof newTime === 'string') {
    newTime = Number(newTime);

    if (isNaN(newTime)) {
      newTime = +new Date(time);
    }
  }

  if (isNaN(newTime)) return '请传入正确的时间格式';
  var now = Date.now();
  var diff = Math.abs(now - newTime);

  if (diff < 1000) {
    return '刚刚';
  }

  var units = {
    年: 1000 * 60 * 60 * 24 * 365,
    个月: 1000 * 60 * 60 * 24 * 30,
    天: 1000 * 60 * 60 * 24,
    小时: 1000 * 60 * 60,
    分钟: 1000 * 60,
    秒: 1000
  };

  for (var unit in units) {
    var timeDiff = Math.floor(diff / units[unit]);

    if (timeDiff > 0) {
      return "".concat(timeDiff).concat(unit).concat(now > newTime ? '前' : '后');
    }
  }

  return '未知时间';
};

function minBy(array, key) {
  var result;

  if (array == null) {
    return result;
  } else if (array != null && array.length === 0) {
    return result;
  }

  var iteratee = function iteratee(value) {
    return value ? value[key] : undefined;
  };

  var computed;

  var _iterator = _createForOfIteratorHelper(array),
      _step;

  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var value = _step.value;
      var current = iteratee(value);

      if (current != null && (computed === undefined ? current === current && !isSymbol(current) : current < computed)) {
        computed = current;
        result = value;
      }
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }

  return result;
}

function getTag(value) {
  var toString = Object.prototype.toString;

  if (value == null) {
    return value === undefined ? '[object Undefined]' : '[object Null]';
  }

  return toString.call(value);
}

function isSymbol(value) {
  var type = _typeof(value);

  return type == 'symbol' || type === 'object' && value != null && getTag(value) === '[object Symbol]';
}

function maxBy(array, key) {
  var result = undefined;

  if (array == null) {
    return result;
  } else if (array.length === 0) {
    return result;
  }

  var iteratee = function iteratee(value) {
    return value ? value[key] : undefined;
  };

  var computed;

  var _iterator = _createForOfIteratorHelper(array),
      _step;

  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var value = _step.value;
      var current = iteratee(value);

      if (current != null && (computed === undefined ? current === current && !isSymbol$1(current) : current > computed)) {
        computed = current;
        result = value;
      }
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }

  return result;
}

function getTag$1(value) {
  var toString = Object.prototype.toString;

  if (value === null) {
    return '[object Null]';
  }

  if (value === undefined) {
    return '[object Undefined]';
  }

  return toString.call(value);
}

function isSymbol$1(value) {
  var type = _typeof(value);

  return type === 'symbol' || type === 'object' && value !== null && getTag$1(value) === '[object Symbol]';
}

var completeIp = function completeIp(ip) {
  try {
    var parts = ip.split('.');

    for (var i = 1; i < parts.length; i++) {
      parts[i] = parts[i].padStart(3, '0');
    }

    return parts.join('.');
  } catch (error) {
    return '请传入正确的 ip 地址！';
  }
};

var capitalizedAmount = function capitalizedAmount(amount) {
  try {
    if (typeof amount === 'string') {
      amount = parseFloat(amount.replaceAll(',', ''));
    }

    if (amount == null || isNaN(amount)) throw new Error('不是有效的金额！');
    var result = '';

    if (amount < 0) {
      result = '欠';
      amount = Math.abs(amount);
    } // 金额不能超过千亿以上


    if (amount >= 10e11) throw new Error('计算金额过大！');
    amount = String(amount.toFixed(2));
    var digits = ['零', '壹', '贰', '叁', '肆', '伍', '陆', '柒', '捌', '玖'];
    var units = ['', '拾', '佰', '仟'];
    var bigUnits = ['', '万', '亿'];
    var decimalUnits = ['角', '分'];
    var amountArray = amount.split('.');
    var integerPart = amountArray[0];
    var decimalPart = amountArray[1]; // 整数部分

    if (integerPart !== '0') {
      integerPart = integerPart.split('').map(Number);
      var levels = integerPart.reverse().reduce(function (prev, item, index) {
        var _prev$;

        var level = (prev === null || prev === void 0 ? void 0 : (_prev$ = prev[0]) === null || _prev$ === void 0 ? void 0 : _prev$.length) < 4 ? prev[0] : [];
        var value = !item ? digits[item] : digits[item] + units[index % 4];
        level.unshift(value);

        if (level.length === 1) {
          prev.unshift(level);
        } else {
          prev[0] = level;
        }

        return prev;
      }, []);
      result += levels.reduce(function (prev, item, index) {
        var _level = bigUnits[levels.length - index - 1];

        var _item = item.join('').replace(/(零)\1+/g, '$1');

        if (_item === '零') {
          _level = '';
          _item = '';
        } else if (_item.endsWith('零')) {
          _item = _item.slice(0, _item.length - 1);
        }

        return prev + _item + _level;
      }, '');
    } else {
      result += '零';
    }

    result += '元'; // 小数部分

    if (decimalPart !== '00') {
      if (result === '零元') result = '';

      for (var i = 0; i < decimalPart.length; i++) {
        var digit = parseInt(decimalPart[i]);

        if (digit !== 0) {
          result += digits[digit] + decimalUnits[i];
        }
      }
    } else {
      result += '整';
    }

    return result;
  } catch (error) {
    return error.message;
  }
};

var dataDesensitizationMap = {
  idCard: function idCard(data) {
    return data.replace(/(\d{2})\d{14}(\w{2})/, '$1' + repeat('*', 14) + '$2');
  },
  phone: function phone(data) {
    return data.replace(/(\d{3})\d{4}(\d{4})/, '$1****$2');
  },
  bankCard: function bankCard(data) {
    return data.replace(/(\d{4})\d{10}(\w{4})/, '$1' + repeat('*', 10) + '$2');
  },
  address: function address(data) {
    return data.replace(/(\S{2})\S*/, '$1' + repeat('*', data.length - 2));
  },
  custom: function custom(data, begin, desensitization) {
    var end = data.length - begin - desensitization;
    var regexp = RegExp("(\\d{" + begin + "})\\d{" + desensitization + "}(\\w{" + end + "})");
    return data.replace(regexp, '$1' + repeat('*', desensitization) + '$2');
  },
  fixPhone: function fixPhone(data) {
    return data.replace(/(\w{3}-)\w*/, '$1' + repeat('*', data.length - 4));
  },
  email: function email(data) {
    return data.replace(/(\w?)(\w+)(\w)(@\w+\.[a-z]+(\.[a-z]+)?)/, '$1****$3$4');
  },
  username: function username(data) {
    return data.replace(/(\S)(\S*)/, '$1' + repeat('*', data.length - 1));
  }
};

var dataDesensitization = function dataDesensitization(des, data) {
  var begin = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;
  var desensitization = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 0;

  if (des === 'custom') {
    return dataDesensitizationMap['custom'](data, begin, desensitization);
  } else if (dataDesensitizationMap[des]) {
    return dataDesensitizationMap[des](data);
  } else {
    throw Error('请输入正确的脱敏类型');
  }
};

/**
 * 定义链表结构
 * @param data
 * @param next
 * @constructor
 */
var Node = function Node(data) {
  _classCallCheck(this, Node);

  this.data = void 0;
  this.next = void 0;
  this.prev = void 0;
  this.data = data;
};

var BothLinkedList = /*#__PURE__*/function () {
  function BothLinkedList() {
    _classCallCheck(this, BothLinkedList);

    this.head = void 0;
    this.tail = void 0;
    this.size = 0;
  }

  _createClass(BothLinkedList, [{
    key: "insertHead",
    value:
    /**
     * 插入元素 采用尾插法
     * @param data
     */
    function insertHead(data) {
      var node = new Node(data);

      if (this.size === 0) {
        this.head = node;
        this.tail = node;
      } else {
        node.prev = this.tail;
        this.tail.next = node;
        this.tail = node;
      }

      this.size++;
    }
    /**
     * 根据索引插入元素
     * @param data
     * @param index
     */

  }, {
    key: "insertIndex",
    value: function insertIndex(data, index) {
      if (index < 0 || index > this.size) {
        throw new Error('要插入的索引已经超过了链表的最大长度');
      }

      var node = new Node(data);
      /**
       * 如果链表为空，直接插入
       */

      if (this.head === null) {
        this.head = node;
        this.tail = node;
      }
      /**
       * 如果 index === 0 则插入的位置是头
       * 如果 index === this.size 则插入的位置是尾
       */


      if (index === 0) {
        node.next = this.head;
        this.head.prev = node;
        this.head = node;
      } else if (index === this.size) {
        node.prev = this.tail;
        this.tail.next = node;
        this.tail = node;
      } else {
        var current = this.head;

        for (var i = 0; i < index - 1; i++) {
          current = current.next;
        }

        node.prev = current;
        node.next = current.next;
        current.next.prev = node;
        current.next = node;
      }

      this.size++;
    }
    /**
     * 删除元素(根据元素值进行删除)
     */

  }, {
    key: "deleteData",
    value: function deleteData(data) {
      if (this.size === 0) {
        throw Error('双向链表为空，不能删除');
      }

      var current = this.head;

      while (current) {
        if (current.data === data) {
          if (this.size === 1) {
            this.head = undefined;
            this.tail = undefined;
          } else if (current === this.head) {
            this.head = this.head.next;
            this.head.prev = undefined;
          } else if (current === this.tail) {
            this.tail = this.tail.prev;
            this.tail.next = undefined;
          } else {
            current.prev.next = current.next;
            current.next.prev = current.prev;
          }

          this.size--;
          return;
        }

        current = current.next;
      }
    }
    /**
     * 根据元素索引删除
     * @param index
     */

  }, {
    key: "deleteFrom",
    value: function deleteFrom(index) {
      if (index < 0 || index >= this.size) {
        throw new Error('双向链表越界');
      }

      if (this.size === 1) {
        this.head = undefined;
        this.tail = undefined;
      } else if (index === 0) {
        this.head = this.head.next;
        this.head.prev = undefined;
      } else if (index === this.size - 1) {
        this.tail = this.tail.prev;
        this.tail.next = undefined;
      } else {
        var current = this.head;

        for (var i = 0; i < index; i++) {
          current = current.next;
        }

        current.prev.next = current.next;
        current.next.prev = current.prev;
      }

      this.size--;
    }
    /**
     * 正向遍历
     */

  }, {
    key: "getHead",
    value: function getHead() {
      return this.head;
    }
    /**
     * 反向遍历
     */

  }, {
    key: "getTail",
    value: function getTail() {
      return this.tail;
    }
    /**
     * 获取链表长度
     */

  }, {
    key: "getSize",
    value: function getSize() {
      return this.size;
    }
    /**
     * 更具索引，获取元素值
     * @param index
     */

  }, {
    key: "getData",
    value: function getData(index) {
      if (index < 0 || index >= this.size) {
        throw Error('要查找的索引已经超过了链表的最大长度');
      }

      var current = this.head;

      for (var i = 0; i < index; i++) {
        current = current.next;
      }

      return current.data;
    }
  }]);

  return BothLinkedList;
}();

var XCookie = /*#__PURE__*/function () {
  function XCookie() {
    _classCallCheck(this, XCookie);

    this.initOption = void 0;
    this.initOption = {
      path: './'
    };
  }

  _createClass(XCookie, [{
    key: "checkEnviroment",
    value: function checkEnviroment() {
      if (typeof document === 'undefined') {
        throw new Error('Document object does not exist, please check enviroment.');
      }
    }
  }, {
    key: "set",
    value: function set(key, value) {
      var option = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
      this.checkEnviroment();
      option = Object.assign({}, this.initOption, option);
      var _option = option,
          path = _option.path,
          domain = _option.domain,
          expires = _option.expires,
          maxAge = _option.maxAge;
      var _option2 = option,
          secure = _option2.secure;
      key = encodeURIComponent(key.trim());
      value = encodeURIComponent(value.trim());
      var secureStr = '';
      path = "; path=".concat(path);
      domain && (domain = "; domain=".concat(domain));
      secure && (secureStr = '; secure');

      if (maxAge && +maxAge == maxAge) {
        maxAge = "; max-age=".concat(maxAge);
      } else {
        if (maxAge) {
          console.warn('Invalid max-age.');
        }

        maxAge = '';
      }

      if (expires) {
        switch (expires.constructor) {
          case String:
            expires = '; expires=' + expires;
            break;

          case Number:
            if (expires === Infinity) {
              expires = '; expires=Fri, 31 Dec 9999 23:59:59 GMT';
            } else {
              expires = '; expires=' + new Date(Date.now() + expires * 864e5);
            }

            break;

          case Date:
            expires = '; expires=' + expires.toUTCString();
            break;
        }
      } else {
        expires = '';
      }

      return document.cookie = "".concat(key, "=").concat(value).concat(path).concat(domain).concat(expires).concat(maxAge).concat(secureStr);
    }
  }, {
    key: "get",
    value: function get(key) {
      this.checkEnviroment();

      if (!key) {
        return '';
      }

      var cookieStrArr = document.cookie ? document.cookie.split('; ') : [];

      var _iterator = _createForOfIteratorHelper(cookieStrArr),
          _step;

      try {
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          var cookieStr = _step.value;
          var partCookie = cookieStr.split('=');
          var name = decodeURIComponent(partCookie[0]),
              value = decodeURIComponent(partCookie.slice(1).join('='));

          if (key === name) {
            return value;
          }
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }

      return '';
    }
  }, {
    key: "remove",
    value: function remove(key) {
      var option = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      this.set(key, '', Object.assign({}, option, {
        expires: -1
      }));
    }
  }, {
    key: "allCookies",
    value: function allCookies() {
      this.checkEnviroment();
      var cookiesObj = {};
      var cookieStrArr = document.cookie ? document.cookie.split('; ') : [];

      var _iterator2 = _createForOfIteratorHelper(cookieStrArr),
          _step2;

      try {
        for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
          var cookieStr = _step2.value;
          var partCookie = cookieStr.split('=');
          var key = decodeURIComponent(partCookie[0]),
              value = decodeURIComponent(partCookie.slice(1).join('='));
          cookiesObj[key] = value;
        }
      } catch (err) {
        _iterator2.e(err);
      } finally {
        _iterator2.f();
      }

      return cookiesObj;
    }
  }]);

  return XCookie;
}();

var index = new XCookie();

/**
 * 将 FormData 转换成 JSON
 * @param data: 为 FormData
 */
var formDataToJson = function formDataToJson(data) {
  var json = {}; // 将表单数据转换为 JSON

  data.forEach(function (val, key) {
    json[key] = val;
  });
  return JSON.stringify(json);
};

/**
 * 定义二叉树的结构
 */
var TreeNode = function TreeNode(data, left, right) {
  _classCallCheck(this, TreeNode);

  this.data = void 0;
  this.left = void 0;
  this.right = void 0;
  this.count = 0;
  this.preOrder = [];
  this.centerOrder = [];
  this.postOrder = [];
  this.data = data;
  this.left = left;
  this.right = right;
};
/**
 * 定义二叉树的操作
 */


var BinaryTree = /*#__PURE__*/function () {
  function BinaryTree() {
    _classCallCheck(this, BinaryTree);

    this.root = void 0;
    this.root = undefined;
  }

  _createClass(BinaryTree, [{
    key: "insert",
    value: function insert(data) {
      var newTreeNode = new TreeNode(data, undefined, undefined);

      if (this.root === undefined) {
        this.root = newTreeNode;
        this.root.count++;
      } else {
        var curNode = this.root;
        var parentNode = undefined;

        while (true) {
          parentNode = curNode;

          if (curNode.data > data) {
            curNode = curNode.left;

            if (!curNode) {
              parentNode.left = newTreeNode;
              this.root.count++;
              break;
            }
          } else if (curNode.data < data) {
            curNode = curNode.right;

            if (!curNode) {
              parentNode.right = newTreeNode;
              this.root.count++;
              break;
            }
          } else if (newTreeNode.data === curNode.data) {
            // 如果给定的数据再次出现，就更新计数值
            curNode.count++;
            break;
          }
        }
      }
    }
    /**
     * 根据元素值 寻找 节点
     * @param data
     */

  }, {
    key: "find",
    value: function find(data) {
      var currNode = this.root;

      while (currNode) {
        if (currNode.data == data) {
          return currNode;
        } else if (data < currNode.data) {
          currNode = currNode.left;
        } else {
          currNode = currNode.right;
        }
      }

      return undefined;
    }
    /**
     * 获取根节点
     */

  }, {
    key: "getRoot",
    value: function getRoot() {
      return this.root;
    }
    /**
     * 删除节点
     */

  }, {
    key: "removeNode",
    value: function removeNode(data) {
      this._removeNode(this.root, data);
    }
  }, {
    key: "_removeNode",
    value: function _removeNode(curNode, data) {
      if (curNode === undefined) {
        return undefined;
      }

      if (curNode.data === data) {
        /**
         * 左子树和右子树都为空，说明是叶子节点，可以直接删除
         */
        if (curNode.left === undefined && curNode.right === undefined) {
          return undefined;
        }

        if (curNode.left === undefined) {
          return curNode.right;
        }

        if (curNode.right === undefined) {
          return curNode.left;
        }
        /**
         * 如果两个节点都存在，则需要寻找右子树上的最小值，因为二叉树从左子树到根节点再到右子树，是从小到大排序
         */


        var tmpNode = this.getMinNode(curNode.right);
        curNode.data = tmpNode.data;
        curNode.right = this._removeNode(curNode.right, tmpNode.data);
        this.root.count--;
        return curNode;
      } else if (curNode.data > data) {
        /**
         * 如果当前节点值 > 需要寻找的 data，则需要在左子树继续寻找
         */
        curNode.left = this._removeNode(curNode.left, data);
        this.root.count--;
        return curNode;
      } else if (curNode.data < data) {
        /**
         * 如果当前节点值 < 需要寻找的 data，则需要在右子树继续寻找
         */
        curNode.right = this._removeNode(curNode.right, data);
        this.root.count--;
        return curNode;
      }
    }
    /**
     * 获取最大节点
     * @param curNode
     */

  }, {
    key: "getMaxNode",
    value: function getMaxNode() {
      var node = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.root;
      var curNode = node;

      while (curNode.right) {
        curNode = curNode.right;
      }

      return curNode;
    }
    /**
     * 获取最小节点
     * @param curNode
     */

  }, {
    key: "getMinNode",
    value: function getMinNode() {
      var node = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.root;
      var curNode = node;

      while (curNode.left) {
        curNode = curNode.left;
      }

      return curNode;
    }
    /**
     * 先序遍历，遍历结束之后，可以通过 bt.getRoot().preOrder 获取遍历以后的元素数组
     * @param root
     */

  }, {
    key: "preOrderTraversal",
    value: function preOrderTraversal(root) {
      if (root === undefined) {
        return undefined;
      }

      this.preOrderTraversal(root.left);
      this.root.preOrder.push(root.data);
      this.preOrderTraversal(root.right);
    }
    /**
     * 中序遍历，遍历结束之后，可以通过 bt.getRoot().centerOrder 获取遍历以后的元素数组
     * @param root
     */

  }, {
    key: "centerOrderTraversal",
    value: function centerOrderTraversal(root) {
      if (root === undefined) {
        return undefined;
      }

      this.root.centerOrder.push(root.data);
      this.centerOrderTraversal(root.left);
      this.centerOrderTraversal(root.right);
    }
    /**
     * 后序遍序遍历，遍历结束之后，可以通过 bt.getRoot().postOrder 获取遍历以后的元素数组
     * @param root
     */

  }, {
    key: "postOrderTraversal",
    value: function postOrderTraversal(root) {
      if (root === undefined) {
        return undefined;
      }

      this.postOrderTraversal(root.left);
      this.postOrderTraversal(root.right);
      this.root.postOrder.push(root.data);
    }
  }]);

  return BinaryTree;
}();

var index$1 = new BinaryTree();

function floatAdd(num1, num2) {
  if (!(isNumber(num1) || isNumber(num2))) {
    console.warn('Please pass in the number type');
    return NaN;
  }

  var r1, r2, m;

  try {
    r1 = num1.toString().split('.')[1].length;
  } catch (error) {
    r1 = 0;
  }

  try {
    r2 = num2.toString().split('.')[1].length;
  } catch (error) {
    r2 = 0;
  }

  m = Math.pow(10, Math.max(r1, r2));
  return (num1 * m + num2 * m) / m;
}

function isNumber(num) {
  return typeof num === 'number' && !isNaN(num);
}

function floatMul(num1, num2) {
  if (!(isNumber$1(num1) || isNumber$1(num2))) {
    console.warn('Please pass in the number type');
    return NaN;
  }

  var m = 0,
      s1 = num1.toString(),
      s2 = num2.toString();

  try {
    m += s1.split('.')[1].length;
  } catch (error) {}

  try {
    m += s2.split('.')[1].length;
  } catch (error) {}

  return Number(s1.replace('.', '')) * Number(s2.replace('.', '')) / Math.pow(10, m);
}

function isNumber$1(num) {
  return typeof num === 'number' && !isNaN(num);
}

function floatDiv(num1, num2) {
  if (!(isNumber$2(num1) || isNumber$2(num2))) {
    console.warn('Please pass in the number type');
    return NaN;
  }

  var m1 = 0,
      m2 = 0,
      s1 = num1.toString(),
      s2 = num2.toString();

  try {
    m1 += s1.split('.')[1].length;
  } catch (error) {}

  try {
    m2 += s2.split('.')[1].length;
  } catch (error) {}

  return floatMul(Number(s1.replace('.', '')) / Number(s2.replace('.', '')), Math.pow(10, m2 - m1));
}

function isNumber$2(num) {
  return typeof num === 'number' && !isNaN(num);
}

function floatSub(num1, num2) {
  if (!(isNumber$3(num1) || isNumber$3(num2))) {
    console.warn('Please pass in the number type');
    return NaN;
  }

  var r1, r2, m, n;

  try {
    r1 = num1.toString().split('.')[1].length;
  } catch (error) {
    r1 = 0;
  }

  try {
    r2 = num2.toString().split('.')[1].length;
  } catch (error) {
    r2 = 0;
  }

  m = Math.pow(10, Math.max(r1, r2));
  n = r1 >= r2 ? r1 : r2;
  return Number(((num1 * m - num2 * m) / m).toFixed(n));
}

function isNumber$3(num) {
  return typeof num === 'number' && !isNaN(num);
}

/**
 * 判断两个对象是否相等
 */
var isObjEqual = function isObjEqual(obj1, obj2) {
  /**
   * 判断类型是否相同
   */
  if (_typeof(obj1) !== _typeof(obj2)) {
    return false;
  }
  /**
   * 如果是基本类型或者 null，直接判断值是否相等
   */


  if (!obj1 || !obj2 || _typeof(obj1) !== 'object') {
    return obj1 === obj2;
  }
  /**
   * 遍历 obj1 的所有属性
   */


  for (var key in obj1) {
    if (!isObjEqual(obj1[key], obj2[key])) {
      return false;
    }
    /**
     * 如果 obj2 没有 obj1 的这个属性
     */


    if (!Reflect.has(obj2, key)) {
      return false;
    }
  }
  /**
   * 遍历 obj2 的所有属性
   */


  for (var _key in obj2) {
    /**
     * 如果 obj1 没有 obj2 的这个属性
     */
    if (!(_key in obj1)) {
      return false;
    }
  }
  /**
   * 所有属性都相等，则两个对象相等
   */


  return true;
};

var objConst = '[object Object]';
/**
 * 计算数组中重复元素以及重复元素的次数
 * @param data
 */

var arrayRepeat = function arrayRepeat(data) {
  var res = {};
  var flag = true;
  data.forEach(function (ele) {
    var type = Object.prototype.toString.call(ele);
    var element = ele;

    if (type === '[object String]') {
      element = JSON.stringify(ele);
    } else if (type === objConst) {
      for (var key in res) {
        var resKey = JSON.parse(key);
        var resType = Object.prototype.toString.call(resKey);

        if (resType === objConst) {
          element = isObjEqual(ele, resKey) ? key : JSON.stringify(ele);
          flag = false;
        }
      }

      if (flag) {
        element = JSON.stringify(ele);
      }
    }

    var count = res[element];

    if (count === undefined || count === null) {
      res[element] = 1;
    } else {
      res[element] = count + 1;
    }
  });
  return res;
};

/*
 * @Author: bigFace2019 599069310@qq.com
 * @Date: 2023-04-11 19:40:55
 * @LastEditors: bigFace2019 599069310@qq.com
 * @LastEditTime: 2023-05-13 11:28:06
 * @FilePath: \xijs\src\getDom\index.ts
 * @Description:计算字符串中重复字符以及重复字符的次数(用对象的方法，key为字符，value为个数，然后遍历对象)
 */

function getRepeatCharacter(str) {
  var obj = {};

  for (var i = 0; i < str.length; i++) {
    var char = str.charAt(i);

    if (obj[char]) {
      obj[char]++;
    } else {
      obj[char] = 1;
    }
  }

  return obj;
}

function calculateMemory(value) {
  var unit = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'B';
  var bytes = 0;

  function sizeOf(obj) {
    if (obj !== null && obj !== undefined) {
      switch (_typeof(obj)) {
        case 'number':
          bytes += 8;
          break;

        case 'string':
          bytes += obj.length * 2;
          break;

        case 'boolean':
          bytes += 4;
          break;

        case 'object':
          if (Array.isArray(obj)) {
            for (var i = 0; i < obj.length; i++) {
              sizeOf(obj[i]);
            }
          } else if (obj !== null) {
            for (var key in obj) {
              if (Object.prototype.hasOwnProperty.call(obj, key)) {
                sizeOf(obj[key]);
              }
            }
          }

          break;

        case 'function':
          bytes += obj.toString().length;
          break;
      }
    }

    return bytes;
  }

  var bytesCount = sizeOf(value);

  switch (unit.toUpperCase()) {
    case 'B':
      return "".concat(bytesCount, " B");

    case 'KB':
      return "".concat((bytesCount / 1024).toFixed(2), " KB");

    case 'MB':
      return "".concat((bytesCount / (1024 * 1024)).toFixed(2), " MB");

    default:
      return "".concat(bytesCount, " B");
  }
}

/**
 * 计算指定字符在字符串中出现的次数
 */
var designateRepeat = function designateRepeat(data, ch) {
  return _toConsumableArray(data).filter(function (item) {
    return item === ch;
  }).length;
};

var Larger;

(function (Larger) {
  Larger[Larger["FALSE"] = 0] = "FALSE";
  Larger[Larger["TRUE"] = 1] = "TRUE";
})(Larger || (Larger = {}));

var PriorityQueue = /*#__PURE__*/function () {
  function PriorityQueue(fn) {
    _classCallCheck(this, PriorityQueue);

    this.pq = void 0;
    this._size = void 0;
    this.compare = void 0;
    this.pq = [-1];
    this._size = 0;
    this.compare = fn || this.larger;
  }

  _createClass(PriorityQueue, [{
    key: "parent",
    value: function parent(cur) {
      return cur >> 1;
    }
  }, {
    key: "left",
    value: function left(cur) {
      return cur * 2;
    }
  }, {
    key: "right",
    value: function right(cur) {
      return cur * 2 + 1;
    }
  }, {
    key: "larger",
    value: function larger(a, b) {
      return a > b ? Larger.TRUE : Larger.FALSE;
    }
  }, {
    key: "swim",
    value: function swim(i) {
      var parentIndex = this.parent(i);

      while (i > 1 && this.compare(this.pq[parentIndex], this.pq[i]) > 0) {
        // 与父元素交换
        var _ref = [this.pq[i], this.pq[parentIndex]];
        this.pq[parentIndex] = _ref[0];
        this.pq[i] = _ref[1];
        i = parentIndex;
        parentIndex = this.parent(i);
      }
    }
  }, {
    key: "sink",
    value: function sink(i) {
      while (this.left(i) <= this._size) {
        var bestIndex = this.left(i); // 右边的子元素比左边的子元素小或者大

        if (this.right(i) <= this._size && this.compare(this.pq[bestIndex], this.pq[this.right(i)]) > 0) {
          bestIndex = this.right(i);
        }

        if (this.compare(this.pq[bestIndex], this.pq[i]) > 0) break;
        var _ref2 = [this.pq[i], this.pq[bestIndex]];
        this.pq[bestIndex] = _ref2[0];
        this.pq[i] = _ref2[1];
        i = bestIndex;
      }
    }
  }, {
    key: "enqueue",
    value: function enqueue(ele) {
      this.pq[++this._size] = ele;
      this.swim(this._size);
    }
  }, {
    key: "dequeue",
    value: function dequeue() {
      if (this._size === 0) {
        return null;
      }

      var bestVal = this.pq[1];
      var _ref3 = [this.pq[this._size], this.pq[1]];
      this.pq[1] = _ref3[0];
      this.pq[this._size] = _ref3[1];
      this.pq.pop();
      this._size--;
      this.sink(1);
      return bestVal;
    }
  }, {
    key: "getHead",
    value: function getHead() {
      return this._size > 0 ? this.pq[1] : null;
    }
  }, {
    key: "size",
    get: function get() {
      return this._size;
    }
  }]);

  return PriorityQueue;
}();

export { $, BothLinkedList, PriorityQueue, arrayRepeat, arrayToListNode, average, base64, index$1 as binaryTree, bubbleSort, calculateMemory, camelize, capitalizedAmount, charCount, cloneDeep, completeIp, coordinatesInCircle, coordinatesInRect, dataDesensitization, dateCalculate, debounce, designateRepeat, downloadFileWithText, downloadFileWithUrl, factorial, fibonacci, readImg as file2img, floatAdd, floatDiv, floatMul, floatSub, formDataToJson, formatDate, formatNumber, formatPercent, genRandomColor, getDomPageXY, getDomScreenXY, getDomText, getRawType, getRepeatCharacter, getRuntimeEnv, getSelection, hex2rgba, hyCompact, hyphenate, isArray, isEmail, isEmpty, isIdCard, isObjEqual, IsPc as isPc, isPhone, judgePointInCircle, lang, linkListToArray, maxBy, minBy, obj2url, parser, quickSort, random, randomStr, redirect, regex, repeat, rgba2obj, shuffle, sleep, store, sum, throttle, timeCutStr, timeSub, transformArray, transformTree, url2obj, index as xCookie, xss };
